### PR TITLE
Move Array#product to Indexable#cartesian_product

### PR DESCRIPTION
**Pull request by HertzDevil**
This PR defines the following functions, which can all take any number of `Indexable`s, in a similar vein to `Enumerable#zip`:

* `Indexable#cartesian_product(*others : Indexable)`
* `Indexable.cartesian_product(indexables : Indexable(Indexable))`
* `Indexable#each_product(*others : Indexable, &block)`
* `Indexable.each_product(indexables : Indexable(Indexable), reuse = false, &block)`

As a bonus, external iteration is now supported via the non-block overloads:

* `Indexable#each_product(*others : Indexable)`
* `Indexable.each_product(indexables : Indexable(Indexable), reuse = false)`

Products between same or different `Indexable`s (e.g. `Tuple` × `Array` × `Array` × `Slice`) are now possible. The corresponding `product` / `each_product` methods in `Array` are marked as deprecated and their uses outside specs removed, because of potential confusion with `Enumerable#product` (see #3790, https://github.com/crystal-lang/crystal/pull/3794#issuecomment-269509322).

My decision to exclude `Enumerable` here is deliberate; the current way of iterating through Cartesian products will fail immediately as soon as an infinite collection is encountered, whereas `Indexable` provides a guarantee that `#size` is finite. This unfortunately means #3790 isn't resolved, as `Set` isn't `Indexable`, but this still eliminates the confusion over function naming.

The overloads that take a splat of `Indexable`s return fixed-size `Tuple`s instead of `Array`s for each combination, again similar to `Enumerable#zip`, so they no longer need the `reuse` parameter.

### CODE REVIEW

File: src/indexable.cr [hunk 1]
```diff
@@ -170,6 +170,311 @@ module Indexable(T)
...
+
+  protected def self.pool_slice(pool, size, reuse)
+    if reuse
+      reuse.clear
+      size.times { |i| reuse << pool[i] }
+      reuse
+    else
+      pool[0, size]
+    end
+  end
```
Comment (HertzDevil) [inline src/indexable.cr:347]: These are essentially the same as the ones in `Array`, which are used by `#each_permutation`, `#each_combination`, and `#each_repeated_combination`; it follows that these methods (and their non-enumerating variants) can also be moved to `Indexable` (same goes for `#each_repeated_permutation` which uses the Cartesian product indirectly).

**Comment by asterite** [comment]
@HertzDevil I didn't have time to look at all your contributions, but they are **great**: well explained or documented, and useful. Keep them coming! Thank you! 🎉 🙏

**Comment by straight-shoota** [comment]
This looks really great.

I think `each_product` should be named `each_cartesian_product` to keep the usual naming convention for iterating methods. It's a bit longer but consistency is valueable.

**Comment by asterite** [comment]
I think only those methods that return an Iterator should be prefixed with `each`. If they accept a block they shouldn't.

**Comment by asterite** [comment]
actually, nevermind

**Comment by HertzDevil** [comment]
I specifically avoided `each_cartesian_product` because the Cartesian product technically refers to the entire returned array, not its individual elements.

**Comment by straight-shoota** [comment]
Right. But just keeping "cartesian" out of the name doesn't really change that,
How about `each_cartesian_tuple`?

**Comment by HertzDevil** [comment]
+1 for `each_cartesian`, since the class method variant yields arrays instead of tuples

File: src/indexable.cr [hunk 2]
```diff
@@ -170,6 +170,311 @@ module Indexable(T)
...
+  # Traversal of elements starts from the last `Indexable`.
+  #
+  # `#cartesian_product` is preferred over this class method when the quantity
+  # of *indexables* is known in advance.
+  #
+  # ```
+  # Indexable.cartesian_product([[1, 2, 3], [4, 5]]) # => [[1, 4], [1, 5], [2, 4], [2, 5], [3, 4], [3, 5]]
+  # ```
+  def self.cartesian_product(indexables : Indexable(Indexable))
+    capacity = indexables.product(&.size)
```
Comment (yxhuvud) [inline src/indexable.cr:207]: Do you need to give a more readable error in the empty case?

Comment (HertzDevil) [reply src/indexable.cr:207]: Actually the current code is wrong; if *indexables* is empty then it should return `[[]]` of an appropriate type, it's only when *indexables* is non-empty and contains an empty collection that the method returns `[]`. Passing empty collections to `cartesian_product` should not be an error.

**Comment by Fryguy** [comment]
Curious why you chose Indexable#cartesian_product vs just Indexable#product?  The cartesian bit seems unnecessary, unless there's another kind of product that could be applied to (more-or-less) ordered vectors?

Ruby has `product` and creating something different in crystal is just yet-another-thing that makes them different (and thus complicated to switch between), so I'm not sure it's worth the rename.

**Comment by straight-shoota** [comment]
@Fryguy See #3790

**Comment by straight-shoota** [comment]
I don't understand the last commit. What's the point of yielding an empty array? Doesn't make sense to me. When the input is empty, there should be no output.

**Comment by HertzDevil** [comment]
When there are 0 collections, there is exactly one way to "pick one element from each of the collections", and that is to select nothing at all. In contrast, when there are some collections but any one is empty, there is no way to pick these elements, so nothing is returned in that case.

This is consistent with the fact that the Cartesian product's cardinality is the numerical product of the cardinalities of each (finite) collection, and we already have `[].product == 1`.

Also on further thought `product_each` might be a better name than what we have so far.

**Comment by straight-shoota** [comment]
Sounds good :+1:

`CartesianProductIteratorEmpty` could probably be merged with `CartesianProductIteratorN`, though. The behaviour isn't that much special, it should just need a few checks for `indexables.empty?` but that would save an additional type.

**Comment by Sija** [comment]
`product_each` reads really bad, unlike any other `each` in stdlib. I'd stick to the `each_*` convention.

**Comment by asterite** [comment]
Also note my comment: https://github.com/crystal-lang/crystal/issues/3790#issuecomment-742397201

I don't think we should change the name `product`.

File: src/indexable.cr [hunk 3]
```diff
@@ -170,6 +170,322 @@ module Indexable(T)
...
+  #
+  # * If *reuse* is an `Array`, it will be reused
+  # * If *reuse* is truthy, the method will create a new `Array` and reuse it
+  # * If *reuse* is falsey, no `Array`s will be reused.
+  #
+  # This can be used to prevent many memory allocations when each combination of
+  # interest is to be used in a read-only fashion.
+  def self.product_each(indexables : Indexable(Indexable), reuse = false, &block)
+    lens = indexables.map &.size
+    return if lens.any? &.==(0)
```
Comment (straight-shoota) [inline src/indexable.cr:304]:
```suggestion
    return if lens.any? &.zero?
```

**Review by straight-shoota — CHANGES_REQUESTED** [review]

**Review by straight-shoota — APPROVED** [review]

**Comment by HertzDevil** [comment]
Can we have a final verdict on the method names so that we could get this and `#repeated_permutations` moved in time for 1.0?

**Review by beta-ziliani — APPROVED** [review]
Thanks!

**Pull request merged**